---
title: "Metagenomics Workshop Report"
date: "November 8, 2017"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  cache=TRUE,
  message=FALSE,
  echo = FALSE)
```

```{r}
library(dplyr)
library(ggplot2)
library(readr)
library(tidyr)
library(reshape2)
library(scales)
library(vegan)
library(qiimer)
library(ape)
```

```{r}
group_colors <- c("#0b45a3", "#bc0909")
```

## Quality Control

```{r message=FALSE}
read_qual <- function (fp) {
  read_tsv(fp) %>%
    extract(
      Samples, c("SampleID", "Direction"), 
      regex="([[:alnum:]]+)_([[:alnum:]]+)") %>%
    mutate(SampleID = sub("_R1", "", SampleID)) %>%
    gather(Position, Qual, -SampleID, -Direction) %>%
    mutate(PositionNum = as.integer(sub("-\\d+$", "", Position)))
}
qual_df <- read_qual("deadmice/sunbeam_output/qc/fastqc_quality.tsv")
```

```{r fig.width=6, fig.height=3}
qual_df %>%
  ggplot() +
  geom_step(aes(x=PositionNum, y=Qual, color=SampleID, group=SampleID)) +
  facet_wrap(~ Direction) +
  scale_y_continuous(limits=c(0, 35)) +
  labs(y="Average quality score", x="Position in read", color="") +
  theme_bw()
```


```{r message=FALSE}
read_qc <- function (fp) {
  read_tsv(fp) %>%
    rename(NonHostReads=`false`, HostReads=`true`, SampleID=Samples) %>%
    mutate(study_group = sub("\\d$", "", SampleID))
}
qc_df <- read_qc("deadmice/sunbeam_output/qc/preprocess_summary.tsv")
```

```{r fig.width=3.3, fig.height=3}
qc_df %>%
  mutate(FractionPassing = both_kept / input) %>%
  ggplot() +
  geom_point(aes(x=SampleID, y=FractionPassing, color=study_group)) +
  scale_y_continuous(labels=scales::percent, limits=c(0.99, 1)) +
  scale_color_manual(values = group_colors, guide=F) +
  labs(y="Fraction of reads\npassing quality control", x="") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust = 1, vjust = 0.5))
```


```{r fig.width=3, fig.height=3}
qc_df %>%
  mutate(PercentHost=HostReads / (HostReads + NonHostReads)) %>%
  ggplot() +
  geom_point(aes(x=SampleID, y=PercentHost, color=study_group)) +
  labs(y="Fraction of host reads", x="") +
  scale_y_continuous(labels=scales::percent, limits = c(0, 0.1)) +
  scale_color_manual(values = group_colors, guide=F) +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust = 1, vjust = 0.5))
```


## Taxonomic classification

```{r message=FALSE}
read_kraken <- function (fp) {
  kdf <- read_tsv(fp, skip=1) %>%
    rename(TaxonID=`#OTU ID`, Lineage=`Consensus Lineage`) %>%
    gather(SampleID, Counts, -TaxonID, -Lineage) %>%
    mutate(SampleID = sub("-taxa$", "", SampleID)) %>%
    do(cbind(., split_assignments(.$Lineage))) %>%
    select(-Lineage) %>%
    mutate_at(taxonomic_ranks, funs(sub("[kpcofgs]__", "", .))) %>%
    mutate_at(taxonomic_ranks, funs(ifelse(. %in% "", NA, .))) %>%
    mutate(Species = ifelse(is.na(Species), NA, paste(Genus, Species)))
  return(kdf)
}
get_assignment_cts <- function (df, assignment_rank="Species") {
  df %>%
    mutate(., Assignment = simplify_assignments(
      .[,taxonomic_ranks], rank2 = assignment_rank)) %>%
    group_by(SampleID, Assignment) %>%
    summarize(Counts = sum(Counts)) %>%
    acast(Assignment ~ SampleID)
}
kraken_df <- read_kraken("deadmice/sunbeam_output/classify/kraken/all_samples.tsv")
cts <- get_assignment_cts(kraken_df)
genus_cts <- get_assignment_cts(kraken_df, "Genus")
```

```{r fig.height=6, fig.width=6}
otu_heatmap(
  cts, rownames(cts), threshold = 750,
  cluster_rows=F,
  cellheight=12, cellwidth=12)
```

## Alpha diversity analysis

```{r}
cts_filtered <- apply(cts, 2, function (x) ifelse(x < 10, 0, x))
s <- data_frame(SampleID = colnames(cts)) %>%
  mutate(study_group = sub("\\d$", "", SampleID)) %>%
  mutate(
    Richness10k = vegan::rarefy(t(genus_cts), 10000), 
    InvSimpson = diversity(t(genus_cts), index = "invsimpson"))
```

```{r fig.width=3, fig.height=3}
s %>%
  ggplot() +
  geom_point(aes(x=SampleID, y=Richness10k, color=study_group)) +
  scale_y_continuous(limits=c(0, 500)) +
  scale_color_manual(values = group_colors, guide=F) +
  labs(y="Richness (Number of taxa\nper 10,000 reads)", color="", x="") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust = 1, vjust = 0.5))
wilcox.test(Richness10k ~ study_group, data=s)
```

```{r fig.width=2.75, fig.height=3}
s %>%
  ggplot() +
  geom_point(aes(x=SampleID, y=InvSimpson, color=study_group)) +
  scale_y_continuous(limits=c(0, 15)) +
  scale_color_manual(values = group_colors, guide=F) +
  labs(y="Inverse Simpson diversity\n(evenness)", color="", x="") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, hjust = 1, vjust = 0.5))
wilcox.test(InvSimpson ~ study_group, data=s)
```

## Beta diversity analysis

```{r fig.width=4, fig.height=3, warning=FALSE}
props <- sweep(cts, 2, colSums(cts), "/")
bc <- vegdist(t(props)) %>%
  dist_subset(s$SampleID)
pc <- pcoa(bc)
pcdf <- cbind(s, pc$vectors[s$SampleID,1:3])
pct_var <- round(pc$values$Relative_eig * 100)
pcdf %>%
  ggplot() +
  geom_hline(yintercept=0, linetype="dashed") +
  geom_vline(xintercept=0, linetype="dashed") +
  geom_point(aes(x=Axis.1, y=Axis.2, color=study_group)) +
  scale_color_manual(values = group_colors, guide=F) +
  labs(
    x=paste0("PCoA axis 1 (", pct_var[1], "%)"),
    y=paste0("PCoA axis 2 (", pct_var[2], "%)"),
    color="") +
  theme_bw()
```

Test for difference in group centroids.

```{r}
adonis(bc ~ study_group, s)
```

Test for difference in group dispersion.

```{r}
permutest(betadisper(bc, s$study_group, type="median"))
```

## Metagenomic assembly results

```{r}
sample_contig_fp <- function (sample_id) {
  paste0("deadmice/sunbeam_output/assembly/", sample_id, "_assembly/final-contigs.fa")
}
read_contig_lengths <- function (fp) {
   x <- readLines(fp)
  is_desc <- grepl("^>", x)
  seq_idx <- cumsum(is_desc)
  xn <- nchar(x)
  xn[is_desc] <- 0
  tapply(xn, seq_idx, sum)
}
contig_info <- s %>%
  group_by(SampleID, study_group) %>%
  do(ContigLength = read_contig_lengths(sample_contig_fp(.$SampleID))) %>%
  ungroup() %>%
  unnest(ContigLength)
```

```{r}
get_n50 <- function (seq_lengths) {
  seq_lengths <- sort(seq_lengths, decreasing = T)
  cum_lengths <- cumsum(seq_lengths)
  total_length <- sum(seq_lengths)
  cum_gt_than_half_total <- cum_lengths >= (total_length / 2)
  n50_idx <- min(which(cum_gt_than_half_total))
  n50<-seq_lengths[n50_idx]
}
contig_summary <- contig_info %>%
  group_by(SampleID, study_group) %>%
  summarize(
    Total = sum(ContigLength), 
    Longest = max(ContigLength), 
    N50 = get_n50(ContigLength)) %>%
  ungroup()
```

```{r}
contig_info %>% 
  arrange(SampleID, desc(ContigLength)) %>%
  group_by(SampleID, study_group) %>%
  mutate(Rank = seq_along(ContigLength)) %>%
  ggplot() +
  geom_step(aes(y=ContigLength, x=Rank, color=study_group)) +
  facet_wrap(~ SampleID) +
  scale_y_log10(labels=scales::comma, breaks=c(3e2, 1e3, 3e3, 1e4, 3e4)) +
  scale_x_continuous(breaks=(0:3 * 1e4), labels=c("0", "10k", "20k", "30k")) +
  scale_color_manual(values = group_colors, guide=F) +
  labs(y="Contig length", x="Rank of contig\n(longest to shortest)") +
  theme_bw() +
  theme(panel.spacing.x = unit(1, "lines"))
```


```{r}
contig_info %>% 
  arrange(SampleID, desc(ContigLength)) %>%
  group_by(SampleID, study_group) %>%
  mutate(CumLength = cumsum(ContigLength)) %>%
  mutate(FracTotalLength = CumLength / sum(ContigLength)) %>%
  ggplot() +
  geom_step(aes(y=ContigLength, x=FracTotalLength, color=study_group)) +
  geom_hline(aes(yintercept = N50), linetype="dashed", data=contig_summary) +
  facet_wrap(~ SampleID) +
  scale_y_log10(labels=scales::comma, breaks=c(3e2, 1e3, 3e3, 1e4, 3e4)) +
  scale_x_continuous(labels=scales::percent) +
  scale_color_manual(values = group_colors, guide=F) +
  labs(y="Contig length", x="Percent of total contig length") +
  theme_bw() +
  theme(panel.spacing.x = unit(1, "lines"))
```

```{r}
contig_summary %>% 
  rename(Group=study_group, `Total length`=Total, `Longest contig`=Longest) %>%
  as.data.frame() %>% 
  knitr::kable(format.args = list(big.mark = ','))
```

\pagebreak

## Antibiotic resistance gene database (CARD) results

```{r}
sample_annotation_fp <- function (sample_id) {
  paste0("deadmice/sunbeam_output/annotation/summary/", sample_id, ".tsv")
}
read_annotation <- function (fp) {
  read_tsv(fp) %>%
    filter(!is.na(card))
}
annotation_info <- s %>%
  group_by(SampleID, study_group) %>%
  do(read_annotation(sample_annotation_fp(.$SampleID))) %>%
  ungroup()
annotation_info %>%
  select(SampleID, contig, length, card) %>%
  knitr::kable()
```

